traffic_portal/app/src/common/modules/form/deliveryService/edit/FormEditDeliveryServiceController.js

```diff
@@ -22,6 +22,21 @@ var FormEditDeliveryServiceController = function(deliveryService, origin, type,
 	// extends the FormDeliveryServiceController to inherit common methods
 	angular.extend(this, $controller('FormDeliveryServiceController', { deliveryService: deliveryService, dsCurrent: deliveryService, origin: origin, type: type, types: types, $scope: $scope }));
 
+	var confirmNoOutstandingRequests = function(deliveryService) {
+		deliveryServiceRequestService.getDeliveryServiceRequests()
+		.then(
+			function(dsRequests) {
+				// search all requests for those that are not completed and share the same deliveryservice id
+				angular.forEach(dsRequests, function(value) {
+					if (value.status != 'complete' && value.deliveryService.id == deliveryService.id) {
```

Comment:
you should short circuit this condition by putting this as the first check as it will fail in most instances and then the 2nd check is never evaluated (might get you a little performance for free)

`if (value.deliveryService.id == deliveryService.id && value.status != 'complete') {`

Reply:
also, if you rename the function something like `hasOpenRequest` and check for the statuses we know to be open (draft, submitted or pending), i think your other condition will read a little nicer

`if (options.status.id == $scope.COMPLETE && !hasOpenRequest(deliveryService))`

the reason i mention this is because this ds request code is a bit of a mess already as i'm sure you noticed, so easy to understand naming will make everyone's life easier.

Reply:
so i've been thinking about this a bit more. the way you are performing this check is rather expensive. you are fetching all the DS requests (open or closed) and then looping thru them to determine if there is an open one for this ds when instead this problem could easily be solved by fixing the order of operations. check out the current code:

`				if (options.status.id == $scope.COMPLETE) {
					deliveryServiceService.updateDeliveryService(deliveryService).
						then(
							function() {
								$state.reload(); // reloads all the resolves for the view
								messageModel.setMessages([ { level: 'success', text: 'Delivery Service [ ' + deliveryService.xmlId + ' ] updated' } ], false);
								createDeliveryServiceUpdateRequest(dsRequest, options.comment, true);
							},
							function(fault) {
								$anchorScroll(); // scrolls window to top
								messageModel.setMessages(fault.data.alerts, false);
							}
						);
				} else {
					createDeliveryServiceUpdateRequest(dsRequest, options.comment, false);
				}
`

Reply:
so i've been thinking about this a bit more. the way you are performing this check is rather expensive. you are fetching all the DS requests (open or closed) and then looping thru them to determine if there is an open one for this ds when instead this problem could easily be solved by fixing the order of operations. check out the current code:

```if (options.status.id == $scope.COMPLETE) {
					deliveryServiceService.updateDeliveryService(deliveryService).
						then(
							function() {
								$state.reload(); // reloads all the resolves for the view
								messageModel.setMessages([ { level: 'success', text: 'Delivery Service [ ' + deliveryService.xmlId + ' ] updated' } ], false);
								createDeliveryServiceUpdateRequest(dsRequest, options.comment, true);
							},
							function(fault) {
								$anchorScroll(); // scrolls window to top
								messageModel.setMessages(fault.data.alerts, false);
							}
						);
				} else {
					createDeliveryServiceUpdateRequest(dsRequest, options.comment, false);
				}```

The order is simply backwards. Currently, it tries to update the ds (which succeeds) and THEN it tries to create a ds request (which fails because there is already one open).

So if  you simply flip the order around and only make it update the ds IF the creation of the ds request succeeds, then problem solved.

Reply:
Thanks, that makes things a lot easier